keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1250 @@
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Kubectl
+
+`kubectl drain` command will be changed to create a NodeMaintenance object instead of marking the
+node unschedulable. We will also change the implementation to skip applications that support
+workload migration. This will be detected by observing a `EvacuationRequest` condition on the pod
+and the subsequent appearance of `EvacuationInitiated` condition within a reasonable timeframe (3m).
+At first only deployments with `.spec.strategy.rollingUpdate.maxSurge` value are expected to
+respond to this request. If the cluster doesn't support the NodeMaintenance API, kubectl will
+perform the node drain in a backwards compatible way.
+
+`kubectl cordon` and `kubectl uncordon` commands will be enhanced with a warning that will warn
+the user if a node is made un/schedulable, and it collides with an existing NodeMaintenance object.
+As a consequence the node maintenance controller will reconcile the node back to the old value.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a set of nodes. We will
+include Cordon and Drain toggles to support the following phases of the maintenance:
+1. Planning: this is to let the users know that maintenance will be performed on a particular set
+   of nodes in the future. Configured with `.spec.cordon=false` and `.spec.drain=false`.
+2. Cordon: stop accepting (scheduling) new pods. Configured with `.spec.cordon=true` and
+   `.spec.drain=false`.-
+3. Drain: gives an intent to drain all selected nodes by setting a `EvacuationRequest`
+   condition with `Reason="NodeMaintenance"` on the node's pods. Configured with
+   `.spec.cordon=true` and `.spec.drain=true`.
+4. Drain Complete: all targeted pods have been drained from all the selected nodes. The nodes can
+   be upgraded, restarted, or shut down. The configuration is still kept at `.spec.cordon=true` and
+   `.spec.drain=true`.
+5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.
+   Set `.spec.cordon=false` and `.spec.drain=false` back again.
+
+```golang
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    // +required
+    NodeSelector *v1.NodeSelector
+    // When set to true, cordons all selected nodes by making them unschedulable.
+    Cordon  bool
+    // When set to true, gives an intent to drain all selected nodes by setting
+    // an EvacuationRequest condition on the node's pods.
+    //
+    // Drain cannot be set to true, unless Cordon is also set to true.
+    Drain bool
```

Comment:
Instead of a series of bools, I'd suggest having a set of predefined states you expect. This will allow you to expand the API surface w/o adding more fields.

Reply:
I was planning to ask this in the API review, but yeah maybe it would be better to start with the more advanced approach first.

Reply:
I have added a `stage` field instead and added a section for supported stage transitions.